### ISSUE TITLE
namespace_imports logic is wrong for import(foo, except=c(bar, baz))

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#1397]
We currently assume that elements of `$imports` returned by `parseNamespaceFile()` that are length>1 are `importFrom` directives; actually, they might be `import(., except=)` directives.

For example, here's dplyr ([permalink](https://github.com/tidyverse/dplyr/blob/55dfc1c5b49eae3f1b144ebe7d9fa2c3d79b1835/NAMESPACE#L473-L474)):

```r
parseNamespaceFile("dplyr", "..")$imports[2:3]
# [[1]]
# [[1]][[1]]
# [1] "vctrs"
# [[1]]$except
# [1] "data_frame"
# [[2]]
# [[2]][[1]]
# [1] "R6"
# [[2]][[2]]
# [1] "R6Class"
```

 And indeed:

```r
lintr:::namespace_imports("dplyr")[435,]
#       pkg        fun
# 435 vctrs data_frame
```

**Comment by IndrajeetPatil** [comment]
I think we should use `getNamespaceImports()` here.

Consider `{dm}` package [NAMESPACE](https://github.com/cynkra/dm/blob/e4c9f834ab20671c19fbb9956f1b253337127d8a/NAMESPACE#L351) as example:

```
import(purrr, except = c(list_along, rep_along, invoke, modify, as_function, flatten_dbl, flatten_lgl, flatten_int, flatten_raw, flatten_chr, splice, flatten, prepend, `%@%`, `%||%`))
```

`namespace_imports()` actually completely inverts the `NAMESPACE` specification by saying that the exact functions which are not imported _are_ the ones imported:

``` r
df <- lintr:::namespace_imports(lintr:::find_package("/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/dm/"))

dplyr::filter(df, pkg == "purrr")
#>      pkg         fun
#> 1  purrr  list_along
#> 2  purrr   rep_along
#> 3  purrr      invoke
#> 4  purrr      modify
#> 5  purrr as_function
#> 6  purrr flatten_dbl
#> 7  purrr flatten_lgl
#> 8  purrr flatten_int
#> 9  purrr flatten_raw
#> 10 purrr flatten_chr
#> 11 purrr      splice
#> 12 purrr     flatten
#> 13 purrr     prepend
#> 14 purrr         %@%
#> 15 purrr        %||%
```

But `getNamespaceImports()` correctly recognizes the imports:

```r
ns <- getNamespaceImports("dm")
sort(unname(ns$purrr))
#>   [1] "%>%"                 "accumulate"          "accumulate_right"   
#>   [4] "accumulate2"         "array_branch"        "array_tree"         
#>   [7] "as_mapper"           "as_vector"           "assign_in"          
#>  [10] "at_depth"            "attr_getter"         "auto_browse"        
#>  [13] "chuck"               "compact"             "compose"            
#>  [16] "cross"               "cross_d"             "cross_df"           
#>  [19] "cross_n"             "cross2"              "cross3"             
#>  [22] "detect"              "detect_index"        "discard"            
#>  [25] "done"                "every"               "exec"               
#>  [28] "flatten_df"          "flatten_dfc"         "flatten_dfr"        
#>  [31] "has_element"         "head_while"          "imap"               
#>  [34] "imap_chr"            "imap_dbl"            "imap_dfc"           
#>  [37] "imap_dfr"            "imap_int"            "imap_lgl"           
#>  [40] "imap_raw"            "imodify"             "insistently"        
#>  [43] "invoke_map"          "invoke_map_chr"      "invoke_map_dbl"     
#>  [46] "invoke_map_df"       "invoke_map_dfc"      "invoke_map_dfr"     
#>  [49] "invoke_map_int"      "invoke_map_lgl"      "invoke_map_raw"     
#>  [52] "is_atomic"           "is_bare_atomic"      "is_bare_character"  
#>  [55] "is_bare_double"      "is_bare_integer"     "is_bare_list"       
#>  [58] "is_bare_logical"     "is_bare_numeric"     "is_bare_vector"     
#>  [61] "is_character"        "is_double"           "is_empty"           
#>  [64] "is_formula"          "is_function"         "is_integer"         
#>  [67] "is_list"             "is_logical"          "is_null"            
#>  [70] "is_numeric"          "is_rate"             "is_scalar_atomic"   
#>  [73] "is_scalar_character" "is_scalar_double"    "is_scalar_integer"  
#>  [76] "is_scalar_list"      "is_scalar_logical"   "is_scalar_numeric"  
#>  [79] "is_scalar_vector"    "is_vector"           "iwalk"              
#>  [82] "keep"                "lift"                "lift_dl"            
#>  [85] "lift_dv"             "lift_ld"             "lift_lv"            
#>  [88] "lift_vd"             "lift_vl"             "list_merge"         
#>  [91] "list_modify"         "lmap"                "lmap_at"            
#>  [94] "lmap_if"             "map"                 "map_at"             
#>  [97] "map_call"            "map_chr"             "map_dbl"            
#> [100] "map_depth"           "map_df"              "map_dfc"            
#> [103] "map_dfr"             "map_if"              "map_int"            
#> [106] "map_lgl"             "map_raw"             "map2"               
#> [109] "map2_chr"            "map2_dbl"            "map2_df"            
#> [112] "map2_dfc"            "map2_dfr"            "map2_int"           
#> [115] "map2_lgl"            "map2_raw"            "modify_at"          
#> [118] "modify_depth"        "modify_if"           "modify_in"          
#> [121] "modify2"             "negate"              "none"               
#> [124] "partial"             "pluck"               "pluck<-"            
#> [127] "pmap"                "pmap_chr"            "pmap_dbl"           
#> [130] "pmap_df"             "pmap_dfc"            "pmap_dfr"           
#> [133] "pmap_int"            "pmap_lgl"            "pmap_raw"           
#> [136] "possibly"            "pwalk"               "quietly"            
#> [139] "rate_backoff"        "rate_delay"          "rate_reset"         
#> [142] "rate_sleep"          "rbernoulli"          "rdunif"             
#> [145] "reduce"              "reduce_right"        "reduce2"            
#> [148] "reduce2_right"       "rerun"               "safely"             
#> [151] "set_names"           "simplify"            "simplify_all"       
#> [154] "slowly"              "some"                "tail_while"         
#> [157] "transpose"           "update_list"         "vec_depth"          
#> [160] "walk"                "walk2"               "when"               
#> [163] "zap"
```

<sup>Created on 2022-09-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by IndrajeetPatil** [comment]
Getting the equivalent outputs with the `getNamespaceImports()` approach (need to convert this to use base-R before implementing):

``` r
library(tidyverse)

df <- lintr:::namespace_imports(lintr:::find_package("/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/dm/"))

as_tibble(df)
#> # A tibble: 1,209 × 2
#>    pkg   fun                        
#>    <chr> <chr>                      
#>  1 DBI   dbCanConnect               
#>  2 DBI   SQL                        
#>  3 DBI   .__T__dbGetStatement:DBI   
#>  4 DBI   .__T__dbSendQuery:DBI      
#>  5 DBI   .__T__dbWithTransaction:DBI
#>  6 DBI   dbDisconnect               
#>  7 DBI   .__T__sqlData:DBI          
#>  8 DBI   .__T__dbQuoteString:DBI    
#>  9 DBI   dbGetDBIVersion            
#> 10 DBI   dbListObjects              
#> # … with 1,199 more rows

ns <- getNamespaceImports("dm")
ns[["base"]] <- NULL # exclude base

purrr::map(ns, as_tibble) |> 
  tibble::enframe() |> 
  tidyr::unnest(cols = value) |> 
  dplyr::rename("pkg" = "name", "fun" = "value")
#> # A tibble: 1,460 × 2
#>    pkg   fun                        
#>    <chr> <chr>                      
#>  1 DBI   dbCanConnect               
#>  2 DBI   SQL                        
#>  3 DBI   .__T__dbGetStatement:DBI   
#>  4 DBI   .__T__dbSendQuery:DBI      
#>  5 DBI   .__T__dbWithTransaction:DBI
#>  6 DBI   dbDisconnect               
#>  7 DBI   .__T__sqlData:DBI          
#>  8 DBI   .__T__dbQuoteString:DBI    
#>  9 DBI   dbGetDBIVersion            
#> 10 DBI   dbListObjects              
#> # … with 1,450 more rows
```

<sup>Created on 2022-09-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by MichaelChirico** [comment]
The one downside to `getNamespaceImports()` I can think of is it requires the package to be installed and loadable, whereas our current `parseNamespaceFile()` only requires having the `NAMESPACE` available and parseable.

It also means potential side effects since `loadNamespace()` will be called.

I'm not sure in practice how big the distinction is, and IINM all of the linters where `namespace_imports()` is used _already_ require `loadNamespace()` to work, so maybe it's a moot point.

**Comment by IndrajeetPatil** [comment]
With #1582, the except directives are respected.

For example, `{dm}` currently imports all of `{magrittr}` [except](https://github.com/cynkra/dm/blob/e4c9f834ab20671c19fbb9956f1b253337127d8a/NAMESPACE#L350) one function: 

```r
import(magrittr, except = c(set_names))
```

With #1582, indeed, this function is not present in the returned imports data frame from `namespace_imports()`:

``` r
df <- lintr:::namespace_imports("dm")
dplyr::filter(df, pkg == "magrittr", fun == "set_names")
#> [1] pkg fun
#> <0 rows> (or 0-length row.names)
```

But all other functions are:

```r
dplyr::filter(df, pkg == "magrittr")
#>         pkg                    fun
#> 1  magrittr     multiply_by_matrix
#> 2  magrittr         raise_to_power
#> 3  magrittr                 inset2
#> 4  magrittr         set_attributes
#> 5  magrittr                extract
#> 6  magrittr          divide_by_int
#> 7  magrittr                    %$%
#> 8  magrittr                   %T>%
#> 9  magrittr           is_less_than
#> 10 magrittr           undebug_fseq
#> 11 magrittr            multiply_by
#> 12 magrittr                     or
#> 13 magrittr             use_series
#> 14 magrittr           set_rownames
#> 15 magrittr                 equals
#> 16 magrittr                    not
#> 17 magrittr                    add
#> 18 magrittr                  inset
#> 19 magrittr                    and
#> 20 magrittr             debug_pipe
#> 21 magrittr              divide_by
#> 22 magrittr             debug_fseq
#> 23 magrittr                    mod
#> 24 magrittr                    %>%
#> 25 magrittr                   %<>%
#> 26 magrittr               set_attr
#> 27 magrittr               extract2
#> 28 magrittr           set_colnames
#> 29 magrittr              n'est pas
#> 30 magrittr               subtract
#> 31 magrittr                freduce
#> 32 magrittr                   %!>%
#> 33 magrittr    is_weakly_less_than
#> 34 magrittr                  is_in
#> 35 magrittr is_weakly_greater_than
#> 36 magrittr              set_class
#> 37 magrittr              functions
#> 38 magrittr            pipe_nested
#> 39 magrittr     pipe_eager_lexical
#> 40 magrittr        is_greater_than
#> 41 magrittr      pipe_lazy_masking
```

<sup>Created on 2022-10-03 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Now the only issue is fixing one failing test from the existing test suite, and I am not sure how to fix.

**Comment by MichaelChirico** [comment]
@IndrajeetPatil was this closed by #2200?

**Comment by IndrajeetPatil** [comment]
I am guessing you mean #1582? 

No, it wasn't. That PR was closed after review because there was a regression I couldn't fix, IIRC.
